File: NeuralNetNative/BaseNeuralBlock.cpp
#define NEURALNETNATIVEAPI
#include "BaseNeuralBlock.h"
#include "malloc.h"

namespace NeuralNetNative {
	BaseNeuralBlock::BaseNeuralBlock(int size, BaseNeuralBlock *parent, ActivationFunction *function) {
        Size = size;
		PreviousSize = parent->GetSize();
		Parent = parent;
		Function = function;

		Bias = (float*)_mm_malloc(size*sizeof(float), 32);
        State = (float*)_mm_malloc(size*sizeof(float), 32);
        Net = (float*)_mm_malloc(size*sizeof(float), 32);
		Weights = (float*)_mm_malloc(size*PreviousSize*sizeof(float), 32);
    }

    BaseNeuralBlock::BaseNeuralBlock(int size, int previousSize, ActivationFunction *function) {
        Size = size;
		PreviousSize = previousSize;
		Parent = 0;
		Function = function;

		Bias = (float*)_mm_malloc(size*sizeof(float), 32);
        State = (float*)_mm_malloc(size*sizeof(float), 32);
        Net = (float*)_mm_malloc(size*sizeof(float), 32);
		Weights = (float*)_mm_malloc(size*PreviousSize*sizeof(float), 32);
    }

	BaseNeuralBlock::~BaseNeuralBlock() {
		Size = 0;
		Parent = 0;
		Function = 0;

		_mm_free(Bias);
		_mm_free(State);
		_mm_free(Net);
		_mm_free(Weights);
	}

    float* BaseNeuralBlock::GetState(void) {
        return State;
    }

    float* BaseNeuralBlock::GetNet(void) {
        return Net;
    }

    BaseNeuralBlock* BaseNeuralBlock::GetParent(void) {
        return Parent;
    }

    float* BaseNeuralBlock::GetWeights(void) {
        return Weights;
    }

    void BaseNeuralBlock::SetWeights(float *newWeights) {
        if (Weights != 0) {
			_mm_free(Weights);
		}
		Weights = newWeights;
    }

    float* BaseNeuralBlock::GetBias(void) {
        return Bias;
    }

    ActivationFunction* BaseNeuralBlock::GetActivationFunction(void) {
        return Function;
    }

    int BaseNeuralBlock::GetSize(void) {
        return Size;
    }

	int BaseNeuralBlock::GetPreviousSize(void) {
        return PreviousSize;
    }
}

File: NeuralNet/NeuralNets/BlockType/NeuralNetBlocks/BaseNeuralBlock.cs
using System;
using NeuralNet.ActivationFunctions;

namespace NeuralNet.NeuralNetBlocks {
    [Serializable]
    public abstract class BaseNeuralBlock {
        protected BaseNeuralBlock[] Parents;
        protected float[][] Weights;
        protected float[] Bias;
        protected IActivationFunction ActivationFunction;
        protected float[] State;
        protected float[] Net;

        protected BaseNeuralBlock(int size, BaseNeuralBlock[] parents, IActivationFunction activationFunction) {
            ActivationFunction = activationFunction;
            Bias = new float[size];
            State = new float[size];
            Net = new float[size];
            Parents = new BaseNeuralBlock[parents.Length];
            Weights = new float[parents.Length][];
            for (var i = 0; i < parents.Length; i++) {
                var parent = parents[i];
                Parents[i] = parent;
                Weights[i] = new float[size*parent.Size];
            }
        }

        protected BaseNeuralBlock(int size, int parentSize, IActivationFunction activationFunction) {
            ActivationFunction = activationFunction;
            Bias = new float[size];
            State = new float[size];
            Net = new float[size];
            Parents = new BaseNeuralBlock[] {null};
            Weights = new[] {new float[size*parentSize]};
        }

        public float[] GetState() {
            return State;
        }

        public float[] GetNet() {
            return Net;
        }

        public BaseNeuralBlock[] GetParents() {
            return Parents;
        }

        public float[][] GetWeights() {
            return Weights;
        }

        public float[] GetBias() {
            return Bias;
        }

        public IActivationFunction GetActivationFunction() {
            return ActivationFunction;
        }

		public void SetWeightsFor(int num, float[] newWeights) {
            Weights[num] = newWeights;
        }

		public void SetBias(float[] newBias) {
			Bias = newBias;
		}

        public int Size {
            get { return State.Length; }
        }

    	public IActivationFunction Function {
    		get { return ActivationFunction; }
    	}

        public abstract void Calculate();

        public abstract void Calculate(float[] input);
    }
}
